idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/WrapWithSafeLetCall.kt

```diff
@@ -0,0 +1,277 @@
+                //   foo.bar()
+                // }
+                //
+                // In this case, `foo.bar` is nullable and this fix should change the code to `foo.bar?.let { it() }`. But note that
+                // the PSI structure of the above code is
+                //
+                // - qualifiedExpression: foo.bar()
+                //   - receiver: foo
+                //   - operationTokenNode: .
+                //   - selectorExpression: bar()
+                //     - calleeExpression: bar
+                //     - valueArgumentList: ()
+                //
+                // So we need to explicitly construct the nullable expression text `foo.bar`.
+                val nullableExpressionText =
+                    "${receiverExpression.text}${qualifiedExpression.operationSign.value}${nullableExpression.text}"
+                val newInvokeCallText = "${suggestedVariableName}${callExpression.valueArgumentList?.text ?: ""}"
+                if (qualifiedExpression == expressionToModify) {
+                    expressionToModify.replace(getNewExpression(nullableExpressionText, newInvokeCallText))
+                } else {
+                    qualifiedExpression.replace(factory.createExpression(newInvokeCallText))
+                    expressionToModify.replace(getNewExpression(nullableExpressionText, expressionToModify.text))
+                }
+
+            } else {
+                val nullableExpressionText = when (nullableExpression) {
+                    is KtBinaryExpression, is KtBinaryExpressionWithTypeRHS -> "(${nullableExpression.text})"
+                    else -> nullableExpression.text
+                }
+                nullableExpression.replace(factory.createExpression(suggestedVariableName))
+                expressionToModify.replace(getNewExpression(nullableExpressionText, expressionToModify.text))
+            }
+        }
+    }
+
+    val forUnsafeCall = diagnosticFixFactory<KtFirDiagnostic.UnsafeCall> { diagnostic ->
+        val nullableExpression = diagnostic.receiverExpression
+        createWrapWithSafeLetCallInputForNullableExpressionIfMoreThanImmediateParentIsWrapped(nullableExpression)
+    }
+
+    val forUnsafeImplicitInvokeCall = diagnosticFixFactory<KtFirDiagnostic.UnsafeImplicitInvokeCall> { diagnostic ->
+        val callExpression = diagnostic.psi.parentOfType<KtCallExpression>(withSelf = true) ?: return@diagnosticFixFactory emptyList()
+        createWrapWithSafeLetCallInputForNullableExpression(
+            callExpression.calleeExpression,
+            isImplicitInvokeCallToMemberProperty = isCallingMemberProperty(callExpression)
+        )
+    }
+
+    private fun KtAnalysisSession.isCallingMemberProperty(callExpression: KtCallExpression): Boolean {
+        val calleeName = callExpression.calleeExpression?.text ?: return false
+        val propertySymbol = (callExpression.resolveCall() as? KtVariableWithInvokeFunctionCall)?.target ?: return false
```

Comment:
So, `KtVariableLikeSymbol` has `symbolKind`. Maybe dumb question, but, can we just check if that is `KtSymbolKind.Member` or if that is not `KtSymbolKind.Local`, instead of scope processing below?

Reply:
That would not work for the following since `foo` is a member.

```
object Foo {
    val foo: (String.() -> Unit)? = {}

    fun bar(s: String) {
        s.foo()
    }
}
```

Reply:
Not sure what you meant "that would not work"? Although `foo` in that example is a member, it wouldn't be a variable like symbol?

Reply:
So consider the following instead:
```
interface Str {
    val foo: (() -> Unit)?
}

object Str2 {
    val foo2: (Str.() -> Unit)? = null

    fun bar(s: Str) {
        s.foo()  // s.foo?.let { it() }
        s.foo2() // foo2?.let { s.it() }
    }
}
```
Both `foo` and `foo2` has symbol kind `Member` and both are calls to functional typed properties. But the refactored code should be different due to scoping difference. Let me refactor the code to make the intention clearer.

Reply:
Let me add tests with these two cases side by side.

Reply:
Variable/function renames look better!